src/planner/builder/join_node.go

```diff
@@ -330,40 +314,55 @@ func (j *JoinNode) pushOtherJoin() error {
 		}
 
 		for _, filter := range j.otherJoinOn.right {
-			var parent SelectNode
-			for _, tb := range filter.referTables {
-				tbInfo := j.referTables[tb]
-				if parent == nil {
-					parent = tbInfo.parent
-					continue
-				}
-				if parent != tbInfo.parent {
-					if j.isHint {
-						if parent.Order() < tbInfo.parent.Order() {
-							parent = tbInfo.parent
-						}
-					} else {
-						parent = findLCA(j.Right, parent, tbInfo.parent)
-					}
-				}
-			}
-
-			if mn, ok := parent.(*MergeNode); ok {
-				mn.setWhereFilter(filter)
-			} else {
-				buf := sqlparser.NewTrackedBuffer(nil)
-				filter.expr.Format(buf)
-				return errors.Errorf("unsupported: on.clause.'%s'.in.cross-shard.join", buf.String())
-			}
+			parent := findParent(filter.referTables, j.Right)
+			setFilter(parent, filter)
 		}
 
 		if len(j.otherJoinOn.others) > 0 {
+			j.judgeStrategy(j.otherJoinOn.others)
 			if err := j.pushOtherFilters(j.otherJoinOn.others, true); err != nil {
 				return err
 			}
 		}
 	}
-	return nil
+
+	j.judgeStrategy(j.otherFilter)
+	return j.pushOtherFilters(j.otherFilter, false)
+}
+
+// judgeStrategy judge the join strategy. SortMerge request the cross-shard expression
+// must be a `ComparisonExpr` and its `left|right`'s parent must be `MergeNode`.
+// Otherwise set the join strategy to NestLoop.
+func (j *JoinNode) judgeStrategy(filters []exprInfo) {
+	for _, filter := range filters {
+		if j.Strategy == NestLoop {
+			break
+		}
+		if exp, ok := filter.expr.(*sqlparser.ComparisonExpr); ok {
```

Comment:
The logic Increases software complexity.
if 
the cross-shard expression must be a `ComparisonExpr` and its `left|right`'s parent must be`MergeNode 
else
setNestLoop()

Reply:
done